tests/test_basis_set.py

```diff
@@ -419,3 +422,442 @@ def test_quick_step() -> None:
+        type='GTO',
+        role='orbital',
+        functional_compositions=[
+            AtomCenteredFunction(
+                function_type='p',
+                harmonic_type='spherical',
+                n_primitive=1,
+                exponents=[0.3],
+                contraction_coefficients=[1.0],
+            )
+        ],
+    )
+
+    container = BasisSetContainer(
+        basis_set_components=[acb_oxygen_extra, acb_hydrogen_extra]
+    )
+    numerical_settings.append(container)
+    container.normalize(None, logger)
+
+    d = numerical_settings[0].m_to_dict()
+    assert d['m_def'] == 'nomad_simulations.schema_packages.basis_set.BasisSetContainer'
+    assert len(d['basis_set_components']) == 2
+
+    by_name = {c['basis_set']: c for c in d['basis_set_components']}
+    cO = by_name['user-extra-O']
+    cH = by_name['user-extra-H']
+
+    # oxygen component
+    assert cO['type'] == 'GTO'
+    assert cO['role'] == 'orbital'
+    assert cO['species_scope'] == [o_ref]
+    assert len(cO['functional_compositions']) == 1
+    s0 = cO['functional_compositions'][0]
+    assert s0['function_type'] == 'd'
+    assert s0['n_primitive'] == 1
+    assert np.allclose(s0['exponents'], [0.8])
+    assert np.allclose(s0['contraction_coefficients'], [1.0])
+
+    # hydrogen component (both H atoms)
+    assert cH['type'] == 'GTO'
+    assert cH['role'] == 'orbital'
+    assert cH['species_scope'] == [h1_ref, h2_ref]
+    assert len(cH['functional_compositions']) == 1
+    s1 = cH['functional_compositions'][0]
+    assert s1['function_type'] == 'p'
+    assert s1['n_primitive'] == 1
+    assert np.allclose(s1['exponents'], [0.3])
+    assert np.allclose(s1['contraction_coefficients'], [1.0])
+
+
+def test_mixed_orbital_aux_ecp() -> None:
+    """
+    Mixed atom-centered setup:
+      - Orbital GTO basis for H and O
+      - Auxiliary_scf GTO basis for H and O
+      - ECP only for O
+    Verifies component wiring, roles, species_scope mapping (via m_to_dict), and ECP consistency.
+    """
+    entry = EntryArchive(
+        data=Simulation(
+            model_system=[
+                ModelSystem(
+                    cell=[AtomicCell()],
+                    particle_states=[
+                        AtomsState(
+                            chemical_symbol='H'
+                        ),  # /data/model_system/0/particle_states/0
+                        AtomsState(
+                            chemical_symbol='O'
+                        ),  # /data/model_system/0/particle_states/1
+                    ],
+                )
+            ],
+            model_method=[
+                ModelMethod(
+                    contributions=[
+                        BaseModelMethod(name='kinetic'),
+                        BaseModelMethod(name='electron-ion'),
+                        BaseModelMethod(name='hartree'),
+                    ],
+                    numerical_settings=[],
+                )
+            ],
+        )
+    )
```

Comment:
This looks like a re-usable setup. I would produce it outside of the method, so here you focus in testing.
If you use a similar template often, consider moving it to the pytest `config`.